added documentation for IndexLayout (#1583)
LLNL/RAJA: docs/sphinx/user_guide/feature/view.rst

    }
  }

Index Layout
^^^^^^^^^^^^^  
``RAJA::IndexLayout`` is a layout that uses an index list to access entries of a 
view.  This layout requires an indexing strategy for each dimension, which 

> determines how to map a given input index to an entry within a view.

Reviewer: It seems like the second sentence contradicts the first. Is this a layout that uses index lists or one with options in each dimension?

Author: Good point.  Fixed.

Three indexing strategies are natively supported in RAJA:
DirectIndex, IndexList, and ConditionalIndexList. DirectIndex maps an input

> index to itself, and does not take any arguments as its constructor.  The

Reviewer 2: ```suggestion
index to itself, and does not take any arguments in its constructor.  The

```

IndexList strategy takes a reference to an array containing the list of relevant
indices to map.  With this strategy, a given input index is mapped to the entry
in its list corresponding to that index.  Lastly, the ConditionalIndexStrategy
takes a reference to an array containing an index list. When the referenced index
list is non-null, the ConditionalIndex strategy is equivalent to that
of the IndexList.  Otherwise, if the index list provided to the constructor is a
null pointer, the ConditionalIndexList maps an input index to itself (same as
the DirectIndex strategy).  The ConditionalIndexList strategy is useful if the
index list can potentially be unused for some situations.

A simple illustrative example is shown below::

  int data[2][3];

  for (int i = 0; i < 2; i ++ ) {
    for (int j = 0; j < 3; j ++ ) {
      // fill data[i][j]...
    }
  }

  int index_list[2] = {1,2};

>   auto index_tuple = camp::tuple<RAJA::DirectIndex<>, RAJA::IndexList<>>(

Reviewer 2: ```suggestion
  auto index_tuple = RAJA::tuple<RAJA::DirectIndex<>, RAJA::IndexList<>>(

```

We have a file `RAJA/include/RAJA/util/camp_aliases.hpp` that provides aliases to camp types in the RAJA namespace. This is preferred to make it simpler for users (they don't have to remember whether a type is in RAJA or camp).

                      RAJA::DirectIndex<>(), RAJA::IndexList<>{&index_list[0]});
			   
  auto index_layout = RAJA::make_index_layout(index_tuple, 2, 3);
  auto view = RAJA::make_index_view(&data[0][0], index_layout);

In the above example, a two-dimensional index layout is created with extents 2

> and 3 for the first and second dimension (respectively).  a DirectIndex strategy

Reviewer 2: ```suggestion
and 3 for the first and second dimension, respectively.  A DirectIndex strategy

```

is implemented for the first dimension and an index list with the entries {1,2}
is used for the second dimension.  With this layout, the view created in the
last line above will determine the entry along the first dimension based on the
first input index provided, and the second provided index will be mapped to
that corresponding entry of the index_list for the second dimension.  For
Example, the value stored in view(1,0) will correspond to the value in
data[1][0].  In this case, the DirectIndex strategy maps the first argument in 
the call to view(1,0) to that same value, and the IndexList maps the second 

> argument to the first index in the index_list.

Reviewer: Its easier to understand what equals what if you write it was code.

Reviewer 2: Yes! More concise and unambiguous.

Author: I attempted to clarify this in the last two lines of the code snippet.  Let me know if there's a better way of illustrating this.

Reviewer: See my assert comment.

    }
  }

Index Layout
^^^^^^^^^^^^

> ``RAJA::IndexLayout`` is a layout that uses an indexing strategy for each

Reviewer 2: ```suggestion
``RAJA::IndexLayout`` is a layout that potentially uses a different indexing strategy for each

```
In reality, all RAJA Layouts use an "index strategy" to index into an array of data. The distinguishing characteristic of `IndexLayout` is the fact that the indexing can be different for each dimension, correct?

Author: I think the distinguishing feature should actually be that it can use an index list (or lists) that map input indices to some entry in a view.  I don't really think that the fact that you can assign different indexing 'strategies' for each dimension is the central point here.  I tried rewording this paragraph to convey that.

Reviewer: I agree that is the main thing. Having to set one for each dimension is not something you had to do before though so it seems noteworthy as well.

Author: I re-worded the first paragraph to convey both of the above points.  Let me know if you think more clarification or elaboration on those main points is needed.

dimension to determine how to map a given input index to an entry within a view.

Three indexing strategies are natively supported in RAJA:
DirectIndex, IndexList, and ConditionalIndexList. DirectIndex maps an input
index to itself, and does not take any arguments in its constructor.  The
IndexList strategy takes a reference to an array containing the list of relevant
indices to map.  With this strategy, a given input index is mapped to the entry
in its list corresponding to that index.  Lastly, the ConditionalIndexStrategy
takes a reference to an array containing an index list.  When the referenced

> index list is non-null, the ConditionalIndex strategy is equivalent to that

Author: I'm not sure I'm using the right language here.  Is it proper to say "a reference to an array" to mean a reference to the first element of the array?  Also, is it proper to say that a reference to an array is null or not-null?

Reviewer: It takes a pointer right? You can say it takes a pointer to an array of indices. I would avoid using reference unless you're actually talking about references in a c++ language sense (`int&`).

Reviewer 2: @gberg617 FYI, in valid C++ code, there is no way to create a null reference. Each reference must be assigned to a valid object where the reference is defined. It is possible to encounter a "null" reference in the sense that it does not refer to a valid object, but that means something is wrong with the code.

of the IndexList.  Otherwise, if the index list provided to the constructor is a
null pointer, the ConditionalIndexList maps an input index to itself (same as
the DirectIndex strategy).  The ConditionalIndexList strategy is useful when the
index list is not initialized for some situations.

A simple illustrative example is shown below::

  int data[2][3];

  for (int i = 0; i < 2; i ++ ) {
    for (int j = 0; j < 3; j ++ ) {
      // fill data[i][j]...
    }
  }

  int index_list[2] = {1,2};

  auto index_tuple = RAJA::tuple<RAJA::DirectIndex<>, RAJA::IndexList<>>(
                      RAJA::DirectIndex<>(), RAJA::IndexList<>{&index_list[0]});
			   
  auto index_layout = RAJA::make_index_layout(index_tuple, 2, 3);
  auto view = RAJA::make_index_view(&data[0][0], index_layout);

>   int val0 = view(1,0); // val0 is entry in data[1][1]
>   int val1 = view(1,1); // val1 is entry in data[1][2]

Reviewer: I'm not sure if its more clear with asserts or not but it's a way to show equality in a code example.
```suggestion
  assert( view(1,0) == data[1][1] );
  assert( &view(1,1) == &data[1][2] );
```

In the above example, a two-dimensional index layout is created with extents 2
and 3 for the first and second dimension, respectively.  A DirectIndex strategy
is implemented for the first dimension and an index list with the entries {1,2}
is used for the second dimension.  With this layout, the view created above will
choose the entry along the first dimension based on the first input index
provided, and the second provided index will be mapped to that corresponding
entry of the index_list for the second dimension. Some examples are shown in
the last two lines above.

> .. note::  There is currently no bounds checking implemented for IndexLayout.

Author: Would it be worth noting here that the index list must live in the correct memory space, or is this outside the scope of basic documentation?

Reviewer: That's true for all memory that used with a certain execution backend. Check our other documentation like for forall to see if it is mentioned there.

Reviewer 2: @gberg617 that's probably worth noting in a `note` here.

    }
  }

Index Layout
^^^^^^^^^^^^

``RAJA::IndexLayout`` is a layout that can use an index list to map
input indices to an entry within a view.  Each dimension of the layout
can have its own indexing strategy to determine this mapping.

Three indexing strategies are natively supported in RAJA:

> DirectIndex, IndexList, and ConditionalIndexList. DirectIndex maps an input

Reviewer 2: ```suggestion
``RAJA::DirectIndex``, ``RAJA::IndexList``, and ``RAJA::ConditionalIndexList``. ``DirectIndex`` maps an input

```

To make it clear that the index types are in the RAJA namespace.  After that, you don't need to keep repeating the namespace, but format the type names so they are highlighted similarly to other types in the docs.
